UnitTests/Utilities/GeometryModelRenderer/src/GeometryModelRenderer.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+	Copyright (c) 2021 Technical University of Munich
+	Chair of Computational Modeling and Simulation.
+
+	TUM Open Infra Platform is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License Version 3
+	as published by the Free Software Foundation.
+
+	TUM Open Infra Platform is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "GeometryModelRenderer.h"
+
+struct WorldBuffer
+{
+	buw::Matrix44f viewProjection;
+	buw::Matrix44f projection;
+	buw::Matrix44f view;
+	buw::Vector3f cam;
+	buw::Matrix44f rotation;
+};
+
+GeometryModelRenderer::GeometryModelRenderer(const buw::ReferenceCounted<buw::IRenderSystem>& renderSystem)
+	: renderSystem_{ renderSystem }
+{
+	backBuffer_ = renderSystem_->getBackBufferTarget();
+	backBuffer_->makeCPUReadable();
+
+	width_ = backBuffer_->width();
+	height_ = backBuffer_->height();
+
+	camera_ = buw::makeReferenceCounted<buw::Camera>();
+	cameraController_ = buw::makeReferenceCounted<buw::CameraController>(camera_);
+
+	camera_->frustum() = buw::CameraFrustum(width_, height_, 0.5f, 50000.f, buw::eProjectionType::Perspective);
+	camera_->transformation().offset() = 15;
+	camera_->transformation().yaw() = buw::constantsf::pi_over_4();
+
+	buw::texture2DDescription dsvTD;
+	dsvTD.width = width_;
+	dsvTD.height = height_;
+	dsvTD.format = buw::eTextureFormat::D24_UnsignedNormalizedInt_S8_UnsignedInt;
+	dsvTD.data = nullptr;
+	dsvTD.isCpuReadable = false;
+	dsvTD.useMSAA = renderSystem_->getMSAAEnabled();
+	depthStencilMSAA_ = renderSystem_->createTexture2D(dsvTD, buw::eTextureBindType::DSV);
+
+	viewport_ = renderSystem_->createViewport(buw::viewportDescription(width_, height_));
+
+	buw::constantBufferDescription cbd0;
+	cbd0.sizeInBytes = sizeof(WorldBuffer);
+	cbd0.data = nullptr;
+	worldBuffer_ = renderSystem_->createConstantBuffer(cbd0);
+
+	WorldBuffer world;
+	world.viewProjection = camera_->viewProjectionMatrix();
+	world.projection = camera_->frustum().projectionMatrix();
+	world.view = camera_->transformation().viewMatrix();
+	world.cam = (camera_->transformation().transformationMatrix() * buw::Vector4f(0, 0, 0, 1)).block<3, 1>(0, 0);
+	world.rotation = camera_->transformation().rotationMatrix();
+
+	buw::constantBufferDescription cbd;
+	cbd.sizeInBytes = sizeof(WorldBuffer);
+	cbd.data = &world;
+	worldBuffer_->uploadData(cbd);
```

Comment:
this seems to be duplicated from member funcion `updateWorldData()` - call it, if it is the same.

Reply:
Resolved: 59b6a3f